File: P150370.py
def date_to_days(day):
    y, m, d = map(int, day.split('.'))
    return d + 28 * (m + 12 * y)

def solution(today, terms, privacies):
    answer = []
    y, m, d = map(int, today.split('.'))
    today = date_to_days(today)
    
    term = {}
    for t in terms:
        dst, duration = map(str, t.split())
        term[dst] = 28 * int(duration)
    
    for i in range(len(privacies)):
        privacy = privacies[i]
        date, dst = map(str, privacy.split())
        days = date_to_days(date)
        if days + term[dst] <= today:
            answer.append(i+1)
    
    return answer

File: P150370.java
import java.util.*;

public class P150370 {
    public int getDaysFromDate(String day) {
        String[] target = day.split("[.]");
        return Integer.parseInt(target[2]) + 28*(Integer.parseInt(target[1]) + 12*Integer.parseInt(target[0]));
    }
    public Integer[] solution(String today, String[] terms, String[] privacies) {
        LinkedList<Integer> result = new LinkedList<>();
        int tday = getDaysFromDate(today);
        
        HashMap<String, Integer> term = new HashMap<>();
        
        for (String t:terms) {
            String[] target = t.split(" ");
            String dst = target[0];
            int duration = Integer.parseInt(target[1]);
            term.put(dst, 28*duration);
        }
        
        for (int i=0; i<privacies.length; i++) {
            String privacy = privacies[i];
            String[] target = privacy.split(" ");
            String date = target[0];
            String dst = target[1];
            int days = getDaysFromDate(date);
            
            if (days + term.get(dst) <= tday) result.add(i+1);
        }

        Integer[] answer = new Integer[result.size()];
        result.toArray(answer);
        
        return answer;
    }
}